crates/ra_assists/src/handlers/replace_unwrap_with_match.rs

```diff
@@ -0,0 +1,188 @@
+use std::iter;
+
+use ra_syntax::{
+    ast::{self, make},
+    AstNode,
+};
+use hir::name;
+
+use crate::{Assist, AssistCtx, AssistId};
+use ast::edit::IndentLevel;
+
+// Assist: replace_unwrap_with_match
+//
+// Replaces `unwrap` a `match` expression. Works for Result and Option.
+//
+// ```
+// enum Result<T, E> { Ok(T), Err(E) }
+// fn main() {
+//     let x: Result<i32, i32> = Result::Ok(92);
+//     let y = x.<|>unwrap();
+// }
+// ```
+// ->
+// ```
+// enum Result<T, E> { Ok(T), Err(E) }
+// fn main() {
+//     let x: Result<i32, i32> = Result::Ok(92);
+//     let y = match x {
+//         Ok(a) => a,
+//         _ => unreachable!(),
+//     };
+// }
+// ```
+pub(crate) fn replace_unwrap_with_match(ctx: AssistCtx) -> Option<Assist> {
+    let method_call: ast::MethodCallExpr = ctx.find_node_at_offset()?;
+    let name = method_call.name_ref()?;
+    if name.text() != "unwrap" {
+        return None;
+    }
+    let caller = method_call.expr()?;
+    let ty = ctx.sema.type_of_expr(&caller)?;
+
+    let type_name = ty.as_adt()?.name(ctx.sema.db);
+    if type_name == name![Result] {
+        let ok_path = make::path_unqualified(make::path_segment(make::name_ref("Ok")));
+        let bind_pat = make::bind_pat(make::name("a")).into();
+        let ok_tuple = make::tuple_struct_pat(ok_path, iter::once(bind_pat)).into();
+        let bind_path = make::path_unqualified(make::path_segment(make::name_ref("a")));
+
+        let ok_arm = make::match_arm(iter::once(ok_tuple), make::expr_path(bind_path));
+
+        let unreachable_call = make::unreachable_macro_call().into();
+        let err_arm = make::match_arm(iter::once(make::placeholder_pat().into()), unreachable_call);
+        let match_arm_list = make::match_arm_list(vec![ok_arm, err_arm]);
+        let match_expr = make::expr_match(method_call.expr()?, match_arm_list);
+        let match_expr = IndentLevel::from_node(method_call.syntax()).increase_indent(match_expr);
```

Comment:
Let's move all these `make::` calls inside the `|edit| {` closure. That way, we won't compute the edits unless the user actually selects the assist (once we actually implement that optimization). The `?` needs to stay outside.